kiwi_keg/tools/lib_fileutil.py

```diff
@@ -0,0 +1,129 @@
+# Copyright (c) 2025 SUSE Software Solutions Germany GmbH. All rights reserved.
+#
+# This file is part of keg.
+#
+# keg is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# keg is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with keg. If not, see <http://www.gnu.org/licenses/>
+#
+import filecmp
+import hashlib
+import logging
+import os
+import re
+import subprocess
+import sys
+
+from difflib import Differ
+
+
+def files_equivalent(filename, dir1, dir2, ignore_kiwi_version):
+    path1 = os.path.join(dir1, filename)
+    path2 = os.path.join(dir2, filename)
+
+    if not os.path.exists(path1) or not os.path.exists(path2):
+        return False
+
+    if filename.endswith('.kiwi'):
+        return kiwi_files_equivalent(path1, path2, ignore_kiwi_version)
+
+    if 'tar' in filename.split('.'):
+        return tar_files_equivalent(path1, path2)
+
+    return filecmp.cmp(path1, path2, shallow=False)
+
+
+def kiwi_files_equivalent(old_kiwi, new_kiwi, ignore_version_change):
+    with open(old_kiwi, 'r') as fh:
+        old_kiwi_content = fh.readlines()
+    with open(new_kiwi, 'r') as fh:
+        new_kiwi_content = fh.readlines()
+
+    differ = Differ()
+    diff = differ.compare(old_kiwi_content, new_kiwi_content)
+    ignore_str = 'generated by keg on'
+    if ignore_version_change:
+        ignore_str += r'|<version>[0123456789\.]+</version>'
+    ignore_regex = re.compile(ignore_str)
+
+    for d in diff:
+        if d[0] in '-+' and not ignore_regex.search(d):
+            return False
+
+    return True
+
+
+def tar_files_equivalent(file1, file2):
+    result = subprocess.run(
+        ['tar', 'xf', file1, '-O'],
+        stdout=subprocess.PIPE
+    )
+    sum1 = hashlib.sha256(result.stdout).digest()
+    result = subprocess.run(
+        ['tar', 'xf', file2, '-O'],
+        stdout=subprocess.PIPE
+    )
+    sum2 = hashlib.sha256(result.stdout).digest()
+    return sum1 == sum2
```

Comment:
This is not true for equivalency determination because tar preserves the order in which the files were added. Meaning if I have 3 files in an archive a, b ,c and one archive is created with `tar -czf t1.tar.gz a b c` and the second archive is created with `tar -czf t2.tar.gz  c a b` the tarball are definitely equivalent but because the order of the extracted files is different the computed checksums of the output when these tarballs are extracted would be different and they would be flagged as non equivalent but they are. The output from tar would have to be sorted to get a better chance of determining whether or not they are equivalent via checksums.

Reply:
I think it would be quite rare that the overlay files configuration of an image description was changed in a way that only the order of the files in the tar ball changed but nothing else. In this case we'd have a bit of unnecessary noise in the submit/pull request but I don't think it's worth the effort to avoid that. But happy to add a comment. :)